resources/subnet.rb

```diff
@@ -16,78 +15,143 @@
 # limitations under the License.
 #
 
-default_action :add
-
-property :subnet, String, name_property: true
-property :broadcast, String
-property :netmask, String, required: true
-property :routers, Array, default: []
-property :options, Array, default: []
-property :ddns, String
-property :evals, Array, default: []
-property :key, Hash, default: {}
-property :zones, Array, default: []
-property :conf_dir, String, default: '/etc/dhcp'
-property :next_server, String
-
-include Chef::DSL::Recipe
-
-attr_accessor :pools
-
-def pool(&block)
-  @pools ||= []
-  p = dhcp_pool("#{@name}-pool#{@pools.count}", &block)
-  p.action :nothing
-  @pools << p
-  p
-end
+include Dhcp::Cookbook::Helpers
+
+property :comment, String,
+          description: 'Unparsed comment to add to the configuration file'
+
+property :ip_version, Symbol,
+          equal_to: %i(ipv4 ipv6),
+          default: :ipv4,
+          description: 'The IP version, 4 or 6'
+
+property :conf_dir, String,
+          default: lazy { dhcpd_config_resource_directory(ip_version, declared_type) },
+          description: 'Directory to create configuration file in'
+
+property :cookbook, String,
+          default: 'dhcp',
+          description: 'Template source cookbook'
+
+property :template, String,
+          default: lazy {
+            case ip_version
+            when :ipv4
+              'subnet.conf.erb'
+            when :ipv6
+              'subnet6.conf.erb'
+            end
+          },
+          description: 'Template source file'
+
+property :owner, String,
+          default: lazy { dhcpd_user },
+          description: 'Generated file owner'
+
+property :group, String,
+          default: lazy { dhcpd_group },
+          description: 'Generated file group'
+
+property :mode, String,
+          default: '0640',
+          description: 'Generated file mode'
+
+property :shared_network, [true, false],
+          default: false,
+          description: 'Flag to indicate subnet is used inside a shared_network resource and should not be added to list.conf'
+
+property :subnet, String,
+          required: true,
+          description: 'Subnet network address'
```

Comment:
Why can't this be the name property like it was before? It seems odd to need to specify this in addition to the name of the resource.

Reply:
I'm not so sure on that one `name` is just that a name it doesn't directly relate to the subnet being defined other than the naming of the file, the subnet is an independent property of the name so it feels more 'correct' this way to me 😕. It's the same with all the other resources with the exception of `host` as the name is used to key the host in the configuration.

Reply:
Making it a ``name_property`` wouldn't require it to always be the case and would allow someone to name the file differently than the subnet name if they wanted to. Any issues if you could make this the ``name_propery`` parameter?

Reply:
I didn't think `name_property` worked like that? I was under the impression that it replaced the `name` property but I'll give that a try it's an easy have-it-either-way if i've gotten it wrong. I'll make it conditional either way in that it'll use `name` if `subnet` isn't specified.

Reply:
Well i've learnt something today! 😉 Change incoming.

Reply:
It might make sense to do that in other resources but I haven't had a chance to look at those yet.

Reply:
I think it only make sense (maybe 😉) with subnet, classes and shared_subnets need a name identifier just like hosts and group could be anything so I'm leaving those be.